applications/ecallmgr/src/ecallmgr_fs_transfer.erl

```diff
@@ -22,6 +22,8 @@ attended(Node, UUID, JObj) ->
 
     ReqURI = <<TransferTo/binary, "@", Realm/binary>>,
 
+    ecallmgr_call_command:set_ringback(Node, UUID, JObj),
```

Comment:
@danielfinke code looks good but I am wondering... prior to this change if the ringback was set on during the bridge or early in the call it would persist through the transfer.  As implemented if the ringback is not present on the AMQP request the ringback will be reset to default, removing the potential for any existing ringback on the channel to persist.  That is to say, with this change during a non-SIP transfer ringback is set or reset but can't be maintained.

The same argument could be made with regards to how we (and would continue after this PR) set ringback during bridge.  I would argue that during a bridge you are effectively initializing a channel and specifically with the endpoint parameters so I feel that would align with expectations.  I also feel a mid-call transfer would not intuitively be expected to reset an existing ringback unless explicitly requested.  Using another random parameter as an example, I would not expect the channel to reset its MOH preferences during a transfer unless explicitly part of the request.

Long story shorter, I am thinking that when calling set_ringback during transfer it should negate setting of the default if not provided on the AMQP request.  However, I suspect not setting the default might not correct your issue?

P.S>
However, I have identified why some of these PRs are tricky!  I was going to play with this on my dev box but I don't know exactly what your call scenario is.  I see that you describe the problem as originating a call to predefined endpoints (not 100% sure I understand that terminology, I am assuming that is on-net device as compared to external numbers) then transfer.  That doesn't provide a lot of specifics for me to reproduce the issue ;)  On all the PRs if there was one or more call scenarios that described what steps you take to produce the problem, what Kazoo currently does and what you are expecting that would help tremendously.  A lot of times we are trying to infer the purpose from code which leads to much back and forth ;)  Any time we are required to infer it means we can't quickly appreciate the full issue and impact of the change and so often we are under time constraints so move to the next with every intention to return when we have more time but that delays things.  Thats just the reality I am seeing on some of these, so the lower you can make that initial cognitive load by bootstraping us into your thought process the better! :)

Reply:
@k-anderson OK, let me try to address the purpose a little more clearly.

With regards to "predefined" endpoints, I mean those that are provided, for example, by kz_endpoints:by_owner_id/3 or defined as a loopback, like:
```
                   [{<<"Invite-Format">>, <<"loopback">>}
                   ,{<<"Route">>,  Extension}
                   ,{<<"To-DID">>, Extension}
                   ,{<<"To-Realm">>, kapps_call:account_realm(Call)}
                   ,{<<"Custom-Channel-Vars">>, kz_json:from_list(CCVs)}
                   ])),
```
or similar.

The situation that has the issue I am trying to address is the loopback case. I have not confirmed if it is present with a list of kz_endpoints, but I don't think it would be.

The scenario can be set up like this:

1. Place a call C1 from device D1 to a callflow F1 that answers the call and puts it on hold.
2. Originate a call C2 to device D2 using kapi_resource:publish_originate_req/1, with Resource-Type: originate and Application-Name: bridge (so, a bridge done without kapps_call_command:bridge, such as you would do in the case of call queues because the bridge has to wait a while). Now, you have a leg for D2 that did not have its ringback set in the route_resp (which is what @lazedo was talking about) because it was originated by a list of endpoints, not through callflows, nor was it set by ecallmgr_fs_bridge (because kapps_call_command:bridge wasn't used). D2 has nothing in its FS vars for ringback.
3. D2 executes an ecallmgr_fs_transfer:attended/3 via whatever means to callflow F2. FreeSWITCH att_xfer performs a switch_ivr_originate a executes the loopback provided by the attended fun, resulting in loopbacks A & B and legs to any devices or offnet resources that are bridged to during callflow F2 (let's call those Cx).
4. The agent is getting its audio from C2. That call will not be bridged to the loopbacks (if doing bowout) nor any of Cx, until the far end answers. So, even if the ringback is set/exported during the callflow, it will not be set on C2 until the call is answered, which is too late, obviously.

So my solution was to set it on the call C2 based on settings passed into the attended xfer JObj. This puts the onus on the proc publishing the transfer to compute the desired ringback, which is imperfect, but it's something.

I didn't account, however, for the fact that there might already be ringback settings on the leg doing the attended xfer from any other scenario than my above one. So a modification would be to check somehow if it is already set on the UUID. If so, don't overwrite it. Or maybe that should also still be the responsibility of the proc publishing the transfer. If they send Ringback: undefined, they don't want to set it because it's already set.

I hoooope that is a lot clearer 🙂

Reply:
@k-anderson hey, anything more I could answer to help out with this?